A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,1095 @@
+  // the name of the action to take. The action with the name can be found in
+  // the action list.
+  //
+  // Route contains the action name instead of embedding the Action, so Action
+  // can be shared and reused by multiple routes.
+  message Route {
+    message HeaderMatcher {
+      message Int64Range {
+        int64 start = 1;
+        int64 end = 2;
+      }
+      string name = 1;
+      oneof header_match_specifier {
+        string exact_match = 2;
+        string regex_match = 3;
+        Int64Range range_match = 4;
+        bool present_match = 5;
+        string prefix_match = 6;
+        string suffix_match = 7;
+      }
+      bool invert_match = 8;
+    }
+    message Fraction {
+      // Specifies the numerator. Defaults to 0.
+      uint32 numerator = 1;
+
+      // Specifies the denominator. If the denominator specified is
+      // less than the numerator, the final fractional percentage
+      // is capped at 1 (100%).
+      uint32 denominator = 2;
+    }
+
+    oneof path_matcher {
+      string path = 1;
+      string prefix = 2;
+      string regex = 3;
+    }
+    repeated HeaderMatcher headers = 3;
+
+    // Every time the route is considered for a match, it
+    // must also fall under the percentage of matches
+    // indicated by this field. For some fraction N/D, a
+    // random number in the range [0,D) is selected. If the
+    // number is <= the value of the numerator N, or if the
+    // key is not present, the default value, the router
+    // continues to evaluate the remaining match criteria.
+    Fraction match_fraction = 4;
+
+    string action = 5; // The action name
+  }
+
+  repeated Route route = 1;
+  map<string, Action> action = 2;
+}
+```
+
+The route refers to the action name instead of embedding the action directly,
+so that one action can be shared by multiple routes.
+
+The config must fulfill the following:
+
+*   Routes
+    *   There must be exactly one `path_matcher` rule.
+        *   `path_matcher` is an oneof so this is always true in proto. But json
+            doesn’t have the oneof concept, so a route may have multiple path
+            matchers, and it will be considered invalid.
+    *   The `action` name must be present in the action list.
+    *   Duplicate matchers are allowed.
+        *   Duplicate matcher means that more than one routes have the same
+            matcher, no matter whether the actions are the same or not.
+        *   Because the first match wins, the second matcher will never be used.
+*   Actions
+    *   The actions in list must be referred by the routes
+    *   There must be at least one valid config in `child_policy`
+
+Violating any of these will make the balancer config invalid, and be rejected.
+
+For each action, the config in `child_policy` to be used to create a child LB
+policy will be picked using the same algorithm as [service
+config](https://github.com/grpc/grpc-proto/blob/master/grpc/service_config/service_config.proto#L297).
```

Comment:
Line 297 points to some field inside `LrsLoadBalancingPolicyConfig` . The actual line number should be 377?

Reply:
Updated